File: GettingStarted/UploadingFiles/java17/src/main/java/guides/spring/gs/FileUploadController.java
package guides.spring.gs;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.mvc.method.annotation.MvcUriComponentsBuilder;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import java.io.IOException;
import java.util.stream.Collectors;

@Controller
public class FileUploadController {
    private final StorageService storageService;

    @Autowired
    public FileUploadController(StorageService storageService) {
        this.storageService = storageService;
    }

    @RequestMapping(method = RequestMethod.GET, value = "/")
    public String listUploadedFiles(Model model) throws IOException {
        model.addAttribute(
            "files",
            storageService.loadAll().map(it ->
                MvcUriComponentsBuilder.fromMethodName(
                    FileUploadController.class,
                    "serveFile",
                    it.getFileName().toString()
                ).build().toUri().toString()
            ).collect(Collectors.toList())
        );
        return "uploadForm";
    }

    @RequestMapping(method = RequestMethod.GET, value = "/files/{filename:.+}")
    @ResponseBody
    public ResponseEntity<Resource> serveFile(@PathVariable String filename) {
        Resource file = storageService.loadAsResource(filename);
        return ResponseEntity.ok().header(
            HttpHeaders.CONTENT_DISPOSITION,
            "attachment; filename=\"" + file.getFilename() + "\""
        ).body(file);
    }

    @RequestMapping(method = RequestMethod.POST, value = "/")
    public String handleFileUpload(
        @RequestParam("file") MultipartFile file,
        RedirectAttributes redirectAttributes
    ) {
        storageService.store(file);
        redirectAttributes.addFlashAttribute(
            "message",
            "You successfully uploaded " + file.getOriginalFilename() + "!"
        );
        return "redirect:/";
    }

    @ExceptionHandler(StorageFileNotFoundException.class)
    public ResponseEntity<?> handleStorageFileNotFound(StorageFileNotFoundException exc) {
        return ResponseEntity.notFound().build();
    }
}

File: GettingStarted/UploadingFiles/kotlin/src/main/kotlin/guides/spring/gs/FileUploadController.kt
package guides.spring.gs

import kotlin.streams.asSequence
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.servlet.mvc.method.annotation.MvcUriComponentsBuilder
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.servlet.mvc.support.RedirectAttributes

@Controller
class FileUploadController @Autowired constructor(private val service: StorageService) {
    @RequestMapping(method = [RequestMethod.GET], value = ["/"])
    fun listUploadedFiles(model: Model): String {
        model.addAttribute(
            "files",
            service.loadAll().asSequence().map {
                MvcUriComponentsBuilder.fromMethodName(
                    FileUploadController::class.java,
                    "serveFile",
                    it.fileName.toString()
                ).build().toUri().toString()
            }.toList()
        )
        return "uploadForm"
    }

    @RequestMapping(method = [RequestMethod.GET], value = ["/files/{filename:.+}"])
    @ResponseBody
    fun serveFile(@PathVariable filename: String): ResponseEntity<Resource> {
        val file = service.loadAsResource(filename)
        return ResponseEntity.ok().header(
            HttpHeaders.CONTENT_DISPOSITION,
            "attachment; filename=\"" + file.filename + "\""
        ).body(file)
    }

    @RequestMapping(method = [RequestMethod.POST], value = ["/"])
    fun handleFileUpload(
        @RequestParam("file") file: MultipartFile,
        redirectAttributes: RedirectAttributes
    ): String {
        service.store(file)
        redirectAttributes.addFlashAttribute(
            "message",
            "You successfully uploaded " + file.originalFilename + "!"
        )
        return "redirect:/"
    }

    @ExceptionHandler(StorageFileNotFoundException::class)
    fun handleStorageFileNotFound(ignored: StorageFileNotFoundException): ResponseEntity<Any> {
        return ResponseEntity.notFound().build()
    }
}
